### PR TITLE
Bring up Android device aot on ci

### PR DESCRIPTION
**Pull request by steveisok**
Addresses https://github.com/dotnet/runtime/issues/33381

Some library tests were skipped:
- due to `"Attempting to JIT compile method ... while running in aot-only mode"` (using conditional facts based on `PlatformDetection.IsNotDeviceAOT`; it's considered not to work on both Android and iOS devices) 
- due to a runtime crash occurred on Android AOT (using conditional active issues based on `PlatformDetection.IsNotAndroidAOT`)
- due to the same failures as on Browser AOT (using conditional active issues based on `PlatformDetection.IsNotAndroidAOT `)

Some library test suites were skipped in `tests.proj`.

### CODE REVIEW

**Comment by SamMonoRT** [comment]
cc @fanyang-mono - something we may also need for enabling CI lanes for aot android validation

File: src/libraries/System.Reflection.Metadata/tests/Metadata/TagToTokenTests.cs [hunk 1]
```diff
@@ -95,6 +95,7 @@ private IEnumerable<XxxTag> GetTags()
 
         [Fact]
         [ActiveIssue("https://github.com/dotnet/runtime/issues/50714", typeof(PlatformDetection), nameof(PlatformDetection.IsBuiltWithAggressiveTrimming), nameof(PlatformDetection.IsBrowser))]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/50714", typeof(PlatformDetection), nameof(PlatformDetection.IsAndroidAOT))]
```
Comment (mdh1418) [inline src/libraries/System.Reflection.Metadata/tests/Metadata/TagToTokenTests.cs:98]: Same issue as the wasm one? Or wrong issue linked? If the failure is the exact same and for possibly the same underlying reason, can the issue be updated to include that it is problematic on Android as well? Or maybe its better to create a separate issue if its not clear they're failing for the same reasons.

File: src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/IDispatchImplAttributeTests.cs [hunk 2]
```diff
@@ -13,6 +13,7 @@ public class IDispatchImplAttributeTests
 
         [Theory]
         [ActiveIssue("https://github.com/dotnet/runtime/issues/50717", typeof(PlatformDetection), nameof(PlatformDetection.IsBuiltWithAggressiveTrimming), nameof(PlatformDetection.IsBrowser))]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/50717", typeof(PlatformDetection), nameof(PlatformDetection.IsAndroidAOT))]
```
Comment (mdh1418) [inline src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/IDispatchImplAttributeTests.cs:16]: Same, as for TagToTokenTests.cs, if the failure is the same and known for same underlying reason, can the issue be updated to include that it fails on Android AOT too?

File: src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/SetWin32ContextInIDispatchAttributeTests.cs [hunk 3]
```diff
@@ -12,6 +12,7 @@ public class SetWin32ContextInIDispatchAttributeTests
 
         [Fact]
         [ActiveIssue("https://github.com/dotnet/runtime/issues/50714", typeof(PlatformDetection), nameof(PlatformDetection.IsBuiltWithAggressiveTrimming), nameof(PlatformDetection.IsBrowser))]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/50714", typeof(PlatformDetection), nameof(PlatformDetection.IsAndroidAOT))]
```
Comment (mdh1418) [inline src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/SetWin32ContextInIDispatchAttributeTests.cs:15]: Same, as for TagToTokenTests.cs, if the failure is the same and known for same underlying reason, can the issue be updated to include that it fails on Android AOT too?

File: src/libraries/System.Threading.Tasks.Extensions/tests/ValueTaskTests.cs [hunk 4]
```diff
@@ -1168,6 +1168,7 @@ public void Generic_ToString_Success()
 
         [Theory]
         [ActiveIssue("https://github.com/dotnet/runtime/issues/50722", typeof(PlatformDetection), nameof(PlatformDetection.IsBuiltWithAggressiveTrimming), nameof(PlatformDetection.IsBrowser))]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/50722", typeof(PlatformDetection), nameof(PlatformDetection.IsAndroidAOT))]
```
Comment (mdh1418) [inline src/libraries/System.Threading.Tasks.Extensions/tests/ValueTaskTests.cs:1171]: Same, as for TagToTokenTests.cs, if the failure is the same and known for same underlying reason, can the issue be updated to include that it fails on Android AOT too?

File: src/tasks/AppleAppBuilder/Templates/runtime.m [hunk 5]
```diff
@@ -307,6 +307,7 @@
     mono_jit_set_aot_mode (MONO_AOT_MODE_INTERP);
 #else
     mono_jit_set_aot_mode (MONO_AOT_MODE_FULL);
+    setenv("MONO_AOT_MODE", "aot", TRUE);
```
Comment (mdh1418) [inline src/tasks/AppleAppBuilder/Templates/runtime.m:310]: Can this be done in a separate PR since this is Apple and not Android?

File: src/libraries/tests.proj [hunk 6]
```diff
@@ -177,6 +177,72 @@
     <ProjectExclusions Include="$(RepoRoot)\src\tests\FunctionalTests\Android\Device_Emulator\AOT\Android.Device_Emulator.Aot.Test.csproj" />
   </ItemGroup>
 
+  <ItemGroup Condition="'$(TargetOS)' == 'Android' and '$(EnableTrimming)' == 'true'">
```
Comment (mdh1418) [inline src/libraries/tests.proj:180]: Can we add `'$(RunDisabledAndroidTests)' != 'true'` to the condition here so we can see if these still fail on a draft PR that runs all the tests?

File: src/libraries/tests.proj [hunk 7]
```diff
@@ -177,6 +177,72 @@
...
+    <!-- Crash -->
+    <ProjectExclusions Include="$(MSBuildThisFileDirectory)System.IO.Pipelines/tests/System.IO.Pipelines.Tests.csproj" />
+    <ProjectExclusions Include="$(MSBuildThisFileDirectory)System.Linq.Parallel/tests/System.Linq.Parallel.Tests.csproj" />
+    <ProjectExclusions Include="$(MSBuildThisFileDirectory)System.Memory.Data/tests/System.Memory.Data.Tests.csproj" />
+    <ProjectExclusions Include="$(MSBuildThisFileDirectory)System.Net.Requests/tests/System.Net.Requests.Tests.csproj" />
+
+    <ProjectExclusions Include="$(RepoRoot)\src\tests\FunctionalTests\Android\Device_Emulator\Interpreter\*.Test.csproj" />
+    <ProjectExclusions Include="$(RepoRoot)\src\tests\FunctionalTests\Android\Device_Emulator\JIT\*.Test.csproj" />
+    <ProjectExclusions Include="$(RepoRoot)\src\tests\FunctionalTests\Android\Device_Emulator\InvariantCultureOnlyMode\*.Test.csproj" />
+    <ProjectExclusions Include="$(RepoRoot)\src\tests\FunctionalTests\Android\Device_Emulator\PInvoke\*.Test.csproj" />
```
Comment (mdh1418) [inline src/libraries/tests.proj:243]: Are there any issues tracking these?

File: src/libraries/System.Diagnostics.StackTrace/tests/StackTraceSymbolsTests.cs [hunk 8]
```diff
@@ -10,7 +10,7 @@ namespace System.Diagnostics.SymbolStore.Tests
     public class StackTraceSymbolsTests
     {
         [Fact]
-        [ActiveIssue("https://github.com/dotnet/runtime/issues/51399", TestPlatforms.iOS | TestPlatforms.tvOS | TestPlatforms.MacCatalyst)]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/51399", TestPlatforms.Android | TestPlatforms.iOS | TestPlatforms.tvOS | TestPlatforms.MacCatalyst)]
```
Comment (mdh1418) [inline src/libraries/System.Diagnostics.StackTrace/tests/StackTraceSymbolsTests.cs:13]: Can the issue be updated to include Android?

File: src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs [hunk 9]
```diff
@@ -29,6 +29,9 @@ public static partial class PlatformDetection
         public static bool IsNetBSD => RuntimeInformation.IsOSPlatform(OSPlatform.Create("NETBSD"));
         public static bool IsAndroid => RuntimeInformation.IsOSPlatform(OSPlatform.Create("ANDROID"));
         public static bool IsiOS => RuntimeInformation.IsOSPlatform(OSPlatform.Create("IOS"));
+        public static bool IsAndroidAOT => IsAndroid && IsMonoAOT;
+        public static bool IsNotAndroidAOT => !IsAndroidAOT;
+        public static bool IsNotDeviceAOT => !((IsAndroid || IsiOS) && IsMonoAOT);
```
Comment (mdh1418) [inline src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs:34]: Do we expect this to include tvOS and MacCatalyst ? Is Device specifically Android and iOS (excluding tvOS MacCatalyst)?

File: eng/testing/tests.mobile.targets [hunk 10]
```diff
@@ -26,6 +26,12 @@
     <AdditionalXHarnessArguments Condition="'$(ExpectedExitCode)' != ''">--expected-exit-code $(ExpectedExitCode)</AdditionalXHarnessArguments>
   </PropertyGroup>
 
+  <PropertyGroup Condition="'$(TargetOS)' == 'Android' and '$(EnableTrimming)' == 'true'">
```
Comment (mdh1418) [inline eng/testing/tests.mobile.targets:29]: Could we not just use L12-L23?

File: eng/pipelines/runtime-staging.yml [hunk 11]
```diff
@@ -177,6 +177,45 @@ jobs:
...
+    variables:
+      # map dependencies variables to local variables
+      - name: librariesContainsChange
+        value: $[ dependencies.evaluate_paths.outputs['SetPathVars_libraries.containsChange'] ]
+      - name: monoContainsChange
+        value: $[ dependencies.evaluate_paths.outputs['SetPathVars_mono.containsChange'] ]
+    jobParameters:
+      testGroup: innerloop
+      nameSuffix: AllSubsets_AOT_Mono
+      buildArgs: -s mono+libs+host+packs+libs.tests -c $(_BuildConfig) /p:ArchiveTests=true /p:RunAOTCompilation=true /p:EnableTrimming=true
```
Comment (mdh1418) [inline eng/pipelines/runtime-staging.yml:200]: If we can reuse L12-L23 in tests.mobile.targets, then we can possibly just pass /p:EnableAggressiveTrimming here

**Comment by mdh1418** [comment]
Why is DebuggerSupport enabled for those 3 suites in particular? 
`src/libraries/System.Collections.Immutable/tests/System.Collections.Immutable.Tests.csproj`
`src/libraries/System.Collections.NonGeneric/tests/System.Collections.NonGeneric.Tests.csproj`
`src/libraries/System.Drawing.Primitives/tests/System.Drawing.Primitives.Tests.csproj`

File: src/tasks/AotCompilerTask/MonoAOTCompiler.props [hunk 12]
```diff
@@ -17,7 +17,7 @@
     <MonoAOTCompilerDefaultAotArguments Condition="'$(TargetArchitecture)' == 'arm'" Include="mtriple=armv7-linux-gnueabi-thumbv7s" />
     <MonoAOTCompilerDefaultAotArguments Include="static" />
     <!-- Default trampolines run out for libraries tests -->
-    <MonoAOTCompilerDefaultAotArguments Include="nimt-trampolines=2000" />
+    <MonoAOTCompilerDefaultAotArguments Include="nimt-trampolines=4000" />
```
Comment (mdh1418) [inline src/tasks/AotCompilerTask/MonoAOTCompiler.props:20]: https://github.com/dotnet/runtime/issues/53033#issuecomment-856973365

File: src/libraries/System.Reflection.Metadata/tests/Metadata/TagToTokenTests.cs (hunk 1 above)
Comment (MaximLipnin) [reply src/libraries/System.Reflection.Metadata/tests/Metadata/TagToTokenTests.cs:98]: It's the same, I left a comment in the active issue

File: src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/IDispatchImplAttributeTests.cs (hunk 2 above)
Comment (MaximLipnin) [reply src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/IDispatchImplAttributeTests.cs:16]: same errors, added a comment to the issue.

File: src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/SetWin32ContextInIDispatchAttributeTests.cs (hunk 3 above)
Comment (MaximLipnin) [reply src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/SetWin32ContextInIDispatchAttributeTests.cs:15]: it should be the same

File: src/libraries/System.Threading.Tasks.Extensions/tests/ValueTaskTests.cs (hunk 4 above)
Comment (MaximLipnin) [reply src/libraries/System.Threading.Tasks.Extensions/tests/ValueTaskTests.cs:1171]: same, added a comment

File: src/tasks/AppleAppBuilder/Templates/runtime.m (hunk 5 above)
Comment (MaximLipnin) [reply src/tasks/AppleAppBuilder/Templates/runtime.m:310]: It was discussed a bit earlier that it'd make sense to unite an effort in part of disabling the tests which identically fail on android/ios. I introduced PlatformDetection.IsNotDeviceAOT which uses PlatformDetection.IsMonoAOT relying on this variable.

File: src/libraries/System.Diagnostics.StackTrace/tests/StackTraceSymbolsTests.cs (hunk 8 above)
Comment (MaximLipnin) [reply src/libraries/System.Diagnostics.StackTrace/tests/StackTraceSymbolsTests.cs:13]: Done

File: src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs (hunk 9 above)
Comment (MaximLipnin) [reply src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs:34]: I'm not sure, no issue to include them here or it can be done in a follow-up @steveisok

File: src/libraries/tests.proj (hunk 6 above)
Comment (MaximLipnin) [reply src/libraries/tests.proj]: Done, thanks

File: src/libraries/tests.proj (hunk 7 above)
Comment (MaximLipnin) [reply src/libraries/tests.proj:243]: Not sure if Steve filed any, I'll do

Comment (MaximLipnin) [reply src/libraries/tests.proj]: https://github.com/dotnet/runtime/issues/53922

**Comment by MaximLipnin** [comment]
> Why is DebuggerSupport enabled for those 3 suites in particular?
> `src/libraries/System.Collections.Immutable/tests/System.Collections.Immutable.Tests.csproj`
> `src/libraries/System.Collections.NonGeneric/tests/System.Collections.NonGeneric.Tests.csproj`
> `src/libraries/System.Drawing.Primitives/tests/System.Drawing.Primitives.Tests.csproj`

It's where I observed the related errors on Android AOT

File: src/tasks/AotCompilerTask/MonoAOTCompiler.props (hunk 12 above)
Comment (MaximLipnin) [reply src/tasks/AotCompilerTask/MonoAOTCompiler.props:20]: Not sure about the correctness of this number but it seems that increasing helped to avoid the error of trampoline limitation.

File: eng/testing/tests.mobile.targets (hunk 10 above)
Comment (MaximLipnin) [reply eng/testing/tests.mobile.targets]: updated in https://github.com/dotnet/runtime/pull/50510/commits/e5b2dab5dfc79f4e11f80e3cae3d81a568e56457

File: eng/pipelines/runtime-staging.yml (hunk 11 above)
Comment (MaximLipnin) [reply eng/pipelines/runtime-staging.yml]: updated in https://github.com/dotnet/runtime/pull/50510/commits/e5b2dab5dfc79f4e11f80e3cae3d81a568e56457

File: src/tasks/AotCompilerTask/MonoAOTCompiler.props (hunk 12 above)
Comment (MaximLipnin) [reply src/tasks/AotCompilerTask/MonoAOTCompiler.props:20]: An example of the error: `((null) error) Ran out of trampolines of type 2 in '/data/user/0/net.dot.Microsoft.VisualBasic.Core.Tests/files/System.Private.CoreLib.dll' (limit 2000)`

File: src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs (hunk 9 above)
Comment (mdh1418) [reply src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs:34]: I think naming-wise, `IsNotDeviceAOT` may suggest the platform is `not AOT compiled iOS, Android, or tvOS`, but in this PR, it's `not AOT compiled iOS or Android`. I'm curious if we intend to skip the same test methods if on AOT tvOS (if we plan to have AOT tvOS).

Comment (MaximLipnin) [reply src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs:35]: In order to be consistent with naming, I can add `tvOS` too.

**Comment by MaximLipnin** [comment]
After addressing the latest feedback the number of test failures increased up to ~114. Not sure if it's related to using `EnableAggressiveTrimming` property rather than just  `EnableTrimming` one.

File: src/tasks/AotCompilerTask/MonoAOTCompiler.props (hunk 12 above)
Comment (MaximLipnin) [reply src/tasks/AotCompilerTask/MonoAOTCompiler.props:20]: It still fails even on increased number, for ex.:
```
((null) error) Ran out of trampolines of type 3 in '/data/user/0/net.dot.System.Net.Sockets.Tests/files/System.Private.CoreLib.dll' (limit 4400)
```

Comment (MaximLipnin) [reply src/tasks/AotCompilerTask/MonoAOTCompiler.props:20]: oh, just noticed that it's of `type 3`

**Comment by steveisok** [comment]
This is no longer needed.  Closing.
